### ISSUE TITLE
SMARTS matcher doesn't see aromatic triple bond as aromatic

### ISSUE THREAD

**Issue by adalke** [rdkit/rdkit#2810]
**Configuration:**

- RDKit Version: 2019.09.1
- Operating system: Mac
- Are you using conda? yes
- If you are using conda, which channel did you install the rdkit from? condaforge

**Description:**

The SMARTS pattern "*:*" doesn't match aromatic triple bonds.

RDKit says there are (at least) three examples in ChEMBL:
```
% grep '[a-z]#[a-z]' ~/databases/chembl_24.rdkit.smi
CCc1nc#cn1CC1CC(c2ccccc2)(c2ccccc2)C(=O)O1 CHEMBL423544
CC(C)c1nc#cn1CC1CC(c2ccccc2)(c2ccccc2)C(=O)O1 CHEMBL1191822
CC(C)c1nc#cn1CC1CC(c2ccccc2)(c2ccccc2)C(=O)O1.Cl CHEMBL542971
```
ChEMBL agrees (see https://www.ebi.ac.uk/chembl/compound_report_card/CHEMBL423544/ , for example).

I wanted to match that with SMARTS. First, get the SMILES and show that it's both aromatic and a triple bond:
```
>>> mol = Chem.MolFromSmiles('CCc1nc#cn1CC1CC(c2ccccc2)(c2ccccc2)C(=O)O1')
>>> b = mol.GetBondBetweenAtoms(4, 5)
>>> b.GetIsAromatic()
True
>>> b.GetBondType()
rdkit.Chem.rdchem.BondType.TRIPLE
```
I can match the triple bond:
```
>>> triple_pat = Chem.MolFromSmarts("*#*")
>>> mol.GetSubstructMatches(triple_pat)
((4, 5),)
>>> 
```
But I'm having difficulties matching a triple bond which is also an aromatic bond:
```
>>> triple_aromatic_pat = Chem.MolFromSmarts("*#:*")
>>> mol.GetSubstructMatches(triple_aromatic_pat)
()
```
The problem seems to be that the SMARTS matcher doesn't see that as being aromatic:
```
>>> aromatic_pat = Chem.MolFromSmarts("*:*")
>>> mol.GetSubstructMatches(aromatic_pat)
((2, 3), (2, 6), (3, 4), (5, 6), (11, 12), (11, 16), (12, 13), (13, 14), (14, 15), (15, 16), (17, 18), (17, 22), (18, 19), (19, 20), (20, 21), (21, 22))
>>>
>>> [m for m in mol.GetSubstructMatches(aromatic_pat) if m[0] == 4 or m[1] == 4]
[(3, 4)]
```

**Comment by ptosco** [comment]
Hi Andrew, those three molecules look very much like ChEMBL mistakes.
They all share the same core, and hence the same mistake: `CHEMBL1191822` is like `CHEMBL423544` with one more methyl substituent, whereas `CHEMBL542971` is just the hydrochloride salt of `CHEMBL423544`.
In all of those molecules, the triple bond should just not be there in the SMILES string, i.e. the correct SMILES are
```
CCc1nccn1CC1CC(c2ccccc2)(c2ccccc2)C(=O)O1 CHEMBL423544
CC(C)c1nccn1CC1CC(c2ccccc2)(c2ccccc2)C(=O)O1 CHEMBL1191822
CC(C)c1nccn1CC1CC(c2ccccc2)(c2ccccc2)C(=O)O1.Cl CHEMBL542971
```
That sort of imidazole analogue with a triple bond does not exist, it could at best be an unstable reaction intermediate such as benzyne, if ever, but certainly not a stable molecule.
That said, what you are observing is definitely an inconsistency. The RDKit will allow to put up to 3 triple bonds in a benzene ring, all of which are marked as aromatic, but none of which match the SMARTS aromatic bond pattern.
```
mol = Chem.MolFromSmiles("c1c#cc#cc#1")
Chem.MolToSmiles(mol)
for b in mol.GetBonds():
    print(b.GetBondType(), b.GetIsAromatic())
AROMATIC True
TRIPLE True
AROMATIC True
TRIPLE True
AROMATIC True
TRIPLE True
for a in mol.GetAtoms():
    print(a.GetIsAromatic())
True
True
True
True
True
True
aromatic_pat = Chem.MolFromSmarts("*:*")
mol.GetSubstructMatches(aromatic_pat)
```
While this is an inconsistency and should probably be fixed, a triple bond in an aromatic molecule hardly exists in the real world, and certainly not in those 3 ChEMBL molecules.

**Comment by adalke** [comment]
Frankly, I was surprised to see an aromatic triple bond. I had to do a web search to figure out that they really can exist - I'd never seen one before.

I'm fine with treating this as a low-priority issue.

**Comment by greglandrum** [comment]
There are a couple of things going on here.
First, based on the way the RDKit currently works, the SMARTS `#:` shouldn't ever match anything. Just like `=:`, `-:`, and `-=` should ever match anything: those are all `AND` bond order queries and the RDKit doesn't allow bonds to have more than one order. Bonds also have an "aromatic" flag (so that you can have kekule structures with aromaticity still present), but there's no way to query for that using SMARTS. You can see that here:
```
In [5]: m = Chem.MolFromSmiles('c1ccccc1')                                                                                                                             

In [6]: Chem.Kekulize(m)                                                                                                                                               

In [7]: m.GetSubstructMatches(Chem.MolFromSmarts('c:c'))                                                                                                               
Out[7]: ()

In [9]: m.GetSubstructMatches(Chem.MolFromSmarts('*-*'))                                                                                                                   
Out[9]: ((0, 5), (1, 2), (3, 4))
```
We could change that behavior - so that `:` is not a bond order query but a `isAromatic?` query - but that seem likely to lead to all kinds of trouble (for one thing, it would require picking an arbitrary kekule form for every aromatic system read in from SMILES).

Next: triple bonds and aromaticity
Yes, the RDKit allows this. I'm not sure that it makes sense, and it definitely leads to odd pathologies like what we see here (the "imidazole" ring from your example):
```
In [3]: m = Chem.MolFromSmiles('Cn1c#cnc1')                                                                                                                            

In [4]: m.Debug()                                                                                                                                                      
Atoms:
	0 6 C chg: 0  deg: 1 exp: 1 imp: 3 hyb: 4 arom?: 0 chi: 0
	1 7 N chg: 0  deg: 3 exp: 3 imp: 0 hyb: 3 arom?: 1 chi: 0
	2 6 C chg: 0  deg: 2 exp: 4 imp: 0 hyb: 2 arom?: 1 chi: 0
	3 6 C chg: 0  deg: 2 exp: 4 imp: 0 hyb: 2 arom?: 1 chi: 0
	4 7 N chg: 0  deg: 2 exp: 3 imp: 0 hyb: 3 arom?: 1 chi: 0
	5 6 C chg: 0  deg: 2 exp: 3 imp: 1 hyb: 3 arom?: 1 chi: 0
Bonds:
	0 0->1 order: 1 conj?: 0 aromatic?: 0
	1 1->2 order: 12 conj?: 1 aromatic?: 1
	2 2->3 order: 3 conj?: 1 aromatic?: 1
	3 3->4 order: 12 conj?: 1 aromatic?: 1
	4 4->5 order: 12 conj?: 1 aromatic?: 1
	5 5->1 order: 12 conj?: 1 aromatic?: 1
```
bond order 12 is aromatic in that output.

**Comment by adalke** [comment]
I'm fine for leaving the current behavior as-is. It's such an odd case.

If someone else runs into it, then perhaps it should be documented somehow. But this issue might be documentation enough?

Up until last week, I assumed that a search for "*:*" would return atom pairs which, if re-assembled, would contain only rings. This ended up causing differences between one piece of code of mine, based on that SMARTS match, and another where I used GetIsAromatic() to construct my own ring system finder. And it only appeared because I was cross-checking them using ChEMBL.

After reviewing the Daylight documentation, I think it's clearly implied that "*:#*" isn't supposed to work. It's too bad I don't have access to a Daylight system to see what it does here.

For fun, I tried "[a]:#[a]" against PubChem, which it accepted as a SMARTS query:
```
Percentage of the database searched: 24%. Full database search produces too many hits, please refine your query...
0 results found...
```
